File: clients/intellij-community/src/main/java/dev/cleanroommc/workbench/intellij/community/AtlasRecipeSession.java
package dev.cleanroommc.workbench.intellij.community;

import com.google.gson.JsonObject;
import java.io.*;
import java.nio.ByteBuffer;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.BooleanSupplier;
import static dev.cleanroommc.workbench.intellij.community.AtlasRecipeContract.*;

/** One no-shell Core command, closed with the browser; graph verification is reused. */
final class AtlasRecipeSession implements AutoCloseable {
    private final Process process;
    private final BufferedInputStream input;
    private final ExecutorService readers = Executors.newFixedThreadPool(2, runnable -> {
        Thread thread = new Thread(runnable, "workbench-atlas-session"); thread.setDaemon(true); return thread;
    });
    private final AtomicReference<String> diagnostic = new AtomicReference<>();
    private String graph;
    private int sequence;
    private boolean closed;

    AtlasRecipeSession(CoreLaunch launch, AtlasRecipeClient.GraphPath path, String cwd, BooleanSupplier cancelled) throws IOException {
        ProcessBuilder builder = new ProcessBuilder(launch.command(List.of("atlas", "recipes", "session", path.commandPath())));
        var environment = CommandProcess.scrubbedEnvironment(builder.environment());
        builder.environment().clear(); builder.environment().putAll(environment);
        if (launch.host().equals("native") && cwd != null) builder.directory(new File(cwd));
        process = builder.start(); input = new BufferedInputStream(process.getInputStream());
        readers.submit(() -> {
            try {
                byte[] buffer = new byte[4096]; int count = process.getErrorStream().read(buffer);
                if (count > 0) diagnostic.set(new String(buffer, 0, count, StandardCharsets.UTF_8));
            } catch (IOException ignored) { /* Closing the session closes stderr. */ }
        });
        try {
            JsonObject ready = parseRoot(read(cancelled), "recipe session ready");
            equal(string(ready, "format"), "workbench-atlas-recipe-session-ready-v1", "Unsupported Atlas session");
            require(integer(ready, "schema_version") == 1 && string(ready, "state").equals("ready"), "Atlas session did not open");
            AtlasRecipeContext context = AtlasRecipeContext.parse(ready.getAsJsonObject("context"));
            require(AtlasRecipeContext.sameRoot(context.root(), path.expectedContextRoot()), "Atlas session changed graph path");
            graph = string(ready, "graph_set_id"); equal(graph, context.graphSetId(), "Atlas session graph changed");
        } catch (IOException | RuntimeException error) { close(); throw error; }
    }

    String graph() { return graph; }

    String request(String operation, JsonObject arguments, BooleanSupplier cancelled) throws IOException {
        if (closed) throw new IOException("Atlas session is closed");
        String id = Integer.toString(++sequence);
        JsonObject request = new JsonObject();
        request.addProperty("format", "workbench-atlas-recipe-session-request-v1"); request.addProperty("schema_version", 1);
        request.addProperty("request_id", id); request.addProperty("graph_set_id", graph);
        request.addProperty("operation", operation); request.add("arguments", arguments);
        process.getOutputStream().write((request + "\n").getBytes(StandardCharsets.UTF_8)); process.getOutputStream().flush();
        JsonObject response = parseRoot(read(cancelled), "recipe session response");
        equal(string(response, "format"), "workbench-atlas-recipe-session-response-v1", "Unsupported Atlas response");
        require(integer(response, "schema_version") == 1, "Unsupported Atlas response schema");
        equal(string(response, "graph_set_id"), graph, "Atlas response graph changed");
        equal(string(response, "request_id"), id, "Atlas response request changed");
        if (!string(response, "state").equals("complete")) throw new IOException(string(response, "error"));
        return required(response, "result").toString();
    }

    private String read(BooleanSupplier cancelled) throws IOException {
        Future<String> frame = readers.submit(() -> {
            ByteArrayOutputStream bytes = new ByteArrayOutputStream();
            for (int next; (next = input.read()) != -1;) {
                if (next == '\n') return StandardCharsets.UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT)
                        .onUnmappableCharacter(CodingErrorAction.REPORT).decode(ByteBuffer.wrap(bytes.toByteArray())).toString();
                bytes.write(next);
                if (bytes.size() > MAX_OUTPUT_BYTES) throw new IOException("Atlas page exceeds the client transport boundary");
            }
            throw new IOException("Atlas session ended; reopen and verify the graph");
        });
        try {
            for (;;) {
                if (cancelled.getAsBoolean()) throw new IOException("Atlas query cancelled");
                if (diagnostic.get() != null) throw new IOException(diagnostic.get());
                try { return frame.get(100, TimeUnit.MILLISECONDS); }
                catch (TimeoutException waiting) { /* Poll cancellation and diagnostic channel. */ }
            }
        } catch (ExecutionException error) {
            close(); throw new IOException(error.getCause().getMessage(), error.getCause());
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt(); close(); throw new IOException("Atlas session interrupted", error);
        } catch (IOException error) { close(); throw error; }
        finally { frame.cancel(true); }
    }

    @Override public void close() {
        if (closed) return; closed = true;
        try { process.getOutputStream().close(); } catch (IOException ignored) { }
        process.destroy(); readers.shutdownNow();
        process.onExit().thenRun(() -> { try { input.close(); } catch (IOException ignored) { } });
    }
}

File: clients/vscode/atlasRecipeSession.js
"use strict";
const { spawn } = require("node:child_process");
const { TextDecoder } = require("node:util");
const { scrubbedEnvironment } = require("./coreClient");
const { commandForCoreLaunch, pathForCoreLaunch, resolveCoreLaunch } = require("./coreLaunch");
const { validateAtlasContext, validateAtlasSearch } = require("./developerToolsClient");
const { validateBrowse } = require("./atlasRecipeBrowseValidation");

async function openRecipeSession(executable, root, options = {}) {
  const cancelled = () => Object.assign(new Error("Atlas browsing cancelled"), { name: "AbortError" });
  if (options.signal?.aborted) throw cancelled();
  const launch = resolveCoreLaunch(executable, options);
  const mapped = pathForCoreLaunch(root, launch, "Atlas graph");
  const child = (options.spawn || spawn)(launch.executable, commandForCoreLaunch(launch, ["atlas", "recipes", "session", mapped]), {
    cwd: launch.host === "native" ? options.cwd : undefined,
    env: scrubbedEnvironment(options.environment), shell: false, windowsHide: true,
    stdio: ["pipe", "pipe", "pipe"],
  });
  let pending, failure, buffer = "", sequence = 0, closed = false;
  const decoder = new TextDecoder("utf-8", { fatal: true });
  const abort = () => fail(cancelled());
  const detach = () => options.signal?.removeEventListener("abort", abort);
  function fail(error) {
    if (failure) return;
    failure = error;
    closed = true;
    detach();
    if (pending) { const current = pending; pending = null; current.reject(failure); }
    child.stdin.end(); child.kill();
  }
  function next() {
    if (failure) return Promise.reject(failure);
    if (pending) return Promise.reject(new Error("Atlas session already has a pending request"));
    return new Promise((resolve, reject) => { pending = { resolve, reject }; });
  }
  child.stdout.on("data", chunk => {
    try {
      buffer += decoder.decode(chunk, { stream: true });
      if (Buffer.byteLength(buffer, "utf8") > 48 * 1024 * 1024) throw new Error("Atlas page exceeds the client transport boundary");
      let end;
      while ((end = buffer.indexOf("\n")) >= 0) {
        const line = buffer.slice(0, end); buffer = buffer.slice(end + 1);
        if (!pending) throw new Error("Atlas returned an unsolicited session record");
        const value = JSON.parse(line); const current = pending; pending = null; current.resolve(value);
      }
    } catch (error) { fail(error); }
  });
  child.stderr.on("data", chunk => fail(new Error(`Atlas session failed: ${chunk.toString("utf8").slice(0, 4000)}`)));
  child.on("error", fail);
  child.stdin.on("error", fail);
  child.on("close", code => {
    detach();
    if (!closed || pending || code !== 0) fail(new Error(`Atlas session closed (${code}); reopen and verify the graph`));
  });
  const close = () => {
    if (closed) return;
    closed = true; detach(); child.stdin.end();
    if (pending) fail(cancelled());
  };
  options.signal?.addEventListener("abort", abort, { once: true });
  if (options.signal?.aborted) abort();
  try {
    const ready = await next();
    if (ready.format !== "workbench-atlas-recipe-session-ready-v1" || ready.schema_version !== 1 || ready.state !== "ready") throw new Error("Unsupported Atlas session");
    validateAtlasContext(ready.context, mapped);
    const graph = ready.context.graph_set_id;
    if (!graph || ready.graph_set_id !== graph) throw new Error("Atlas session graph identity changed");
    async function request(operation, arguments_) {
      if (closed) throw new Error("Atlas session is closed");
      const requestId = String(++sequence);
      const response = next();
      child.stdin.write(JSON.stringify({ format: "workbench-atlas-recipe-session-request-v1", schema_version: 1,
        request_id: requestId, graph_set_id: graph, operation, arguments: arguments_ }) + "\n");
      const value = await response;
      if (value.format !== "workbench-atlas-recipe-session-response-v1" || value.schema_version !== 1
          || value.request_id !== requestId || value.graph_set_id !== graph) throw new Error("Atlas session response linkage changed");
      if (value.state !== "complete") throw new Error(typeof value.error === "string" ? value.error : "Atlas request did not complete");
      return value.result;
    }
    return { graph, root: mapped, close,
      search: async query => validateAtlasSearch(await request("search", { query, limit: 50 }), query, 50, mapped),
      browse: async (selection, offset) => {
        return validateBrowse(await request("browse", { selection_id: selection, offset, limit: 50 }), { root: mapped, graph, selection, offset });
      },
    };
  } catch (error) { closed = true; fail(error); throw error; }
}

module.exports = { openRecipeSession };
